src/mapping/mod.rs

```diff
@@ -0,0 +1,355 @@
+use crate::event::{Event, Value};
+
+pub mod parser;
+pub mod query;
+
+use string_cache::DefaultAtom as Atom;
+
+pub type Result<T> = std::result::Result<T, String>;
+
+pub(self) trait Function: Send + core::fmt::Debug {
+    fn apply(&self, target: &mut Event) -> Result<()>;
+}
+
+//------------------------------------------------------------------------------
+
+#[derive(Debug)]
+pub(self) struct Assignment {
+    path: String,
+    function: Box<dyn query::Function>,
+}
+
+impl Assignment {
+    pub(self) fn new(path: String, function: Box<dyn query::Function>) -> Self {
+        Self { path, function }
+    }
+}
+
+impl Function for Assignment {
+    fn apply(&self, target: &mut Event) -> Result<()> {
+        let v = self.function.execute(&target)?;
+        target.as_mut_log().insert(&self.path, v);
+        Ok(())
+    }
+}
+
+//------------------------------------------------------------------------------
+
+#[derive(Debug)]
+pub(self) struct Deletion {
+    // TODO: Switch to String once Event API is cleaned up.
+    path: Atom,
+}
+
+impl Deletion {
+    pub(self) fn new(path: String) -> Self {
+        Self {
+            path: Atom::from(path),
+        }
+    }
+}
+
+impl Function for Deletion {
+    fn apply(&self, target: &mut Event) -> Result<()> {
+        target.as_mut_log().remove(&self.path);
+        Ok(())
+    }
+}
+
+//------------------------------------------------------------------------------
+
+#[derive(Debug)]
+pub(self) struct OnlyFields {
+    paths: Vec<String>,
+}
+
+impl OnlyFields {
+    pub(self) fn new(paths: Vec<String>) -> Self {
+        Self { paths }
+    }
+}
+
+impl Function for OnlyFields {
+    fn apply(&self, target: &mut Event) -> Result<()> {
+        let target_log = target.as_mut_log();
+
+        let keys: Vec<String> = target_log.keys().collect();
```

Comment:
Seems like we could inline this below and avoid the `collect`.

Reply:
I've reduced the allocs by filtering before `collect`, but am I missing a trick somewhere? Didn't think it'd be possible to double borrow `target_log`.

Reply:
Where would the double borrow be? I just meant that what you have here seems equivalent to
```rust
for key in target_log.keys() {
    ...
}
```

Reply:
@lukesteensen I'm also not sure how this could be done without the extra allocations.

At line 74, `let target_log = target.as_mut_log()` borrows mutably, and then without `collect`, we'd be trying to operate on an immutable reference to `target_log` in the for loop.

At least that's what I'm getting from the compiler, while trying to incorporate your proposal.